### PR TITLE
Added a bunch of new tmux settings

### PR DESCRIPTION
**Pull request by wiedzmin**
* more variable prefix bindings
+ couple of seemingly basic options
+ statusbar settings submodule
+ keybindings submodule

### CODE REVIEW

File: modules/programs/firefox.nix [hunk 1]
```diff
@@ -264,6 +270,11 @@ in
           mkIf (profile.settings != {} || profile.extraConfig != "") {
             text = mkUserJs profile.settings profile.extraConfig;
           };
+
+        ".mozilla/firefox/${profile.path}/handlers.json" =
+          mkIf (profile.handlers != {}) {
+            text = builtins.toJSON profile.handlers;
+          };
```
Comment (kalbasit) [inline modules/programs/firefox.nix]: Please move this into a separate PR.

File: modules/programs/tmux.nix [hunk 2]
```diff
@@ -83,6 +227,15 @@ in
 {
   options = {
     programs.tmux = {
+      activeBorderStyle = mkOption {
+        default = defaultShortcut;
```
Comment (kalbasit) [inline modules/programs/tmux.nix]: why the border style is set by default to the shortcut?

Comment (wiedzmin) [reply modules/programs/tmux.nix]: my fault, will fix

File: modules/programs/firefox.nix (hunk 1 above)
Comment (wiedzmin) [reply modules/programs/firefox.nix]: I will remove this altogether, it is not yet to be published.

**Comment by wiedzmin** [comment]
Fixed concerns, please review

File: modules/programs/tmux.nix [hunk 3]
```diff
@@ -8,6 +8,36 @@ let
...
+  bindingsModule = types.submodule {
+    options = {
+      copyMode = mkOption {
+        type = types.attrs;
+        description = "copy-mode bindings.";
+        default = {};
+      };
+      joinPaneMode = mkOption {
+        type = types.attrs;
+        description = "join-pnae mode bindings.";
```
Comment (kalbasit) [inline modules/programs/tmux.nix]:
```suggestion
        description = "join-pane mode bindings.";
```

File: modules/programs/tmux.nix [hunk 4]
```diff
@@ -277,6 +473,79 @@ in
...
+        home.file.".tmux.conf".text = mkOrder 1200 ''
+          ${optionalString (cfg.bindings.copyMode != {})
+                           (builtins.concatStringsSep "\n"
+                             (lib.mapAttrsToList (key: command: ''
+                               bind -T copy-mode ${key} ${command}
+                             '') cfg.bindings.copyMode))}
+          ${optionalString (cfg.bindings.viCopyMode != {})
+                           (builtins.concatStringsSep "\n"
+                             (lib.mapAttrsToList (key: command: ''
+                               bind -T copy-mode-vi ${key} ${command}
```
Comment (kalbasit) [inline modules/programs/tmux.nix]: What if we only have `cfg.bindings.copyMode`, and use `cfg.keyMode` to decide between `copy-mode` and `copy-mode-vi`?

File: modules/programs/tmux.nix [hunk 5]
```diff
@@ -199,10 +382,10 @@ in
 
       shortcut = mkOption {
         default = defaultShortcut;
-        example = "a";
+        example = "C-a";
```
Comment (kalbasit) [inline modules/programs/tmux.nix]: This changes the behavior of the current version. If we proceed with this change, we should probably announce it in the news.

Comment (wiedzmin) [reply modules/programs/tmux.nix]: Thanks, I was in doubt about this.

File: modules/programs/tmux.nix (hunk 4 above)
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Good idea, I think, it will reduce duplication.

File: modules/programs/tmux.nix [hunk 6]
```diff
@@ -150,13 +308,38 @@ in
         description = "Maximum number of lines held in window history.";
       };
 
+      hooks = mkOption {
+        default = { };
+        type = types.attrs;
```
Comment (rycee) [inline modules/programs/tmux.nix]: Perhaps `attrsOf str`? Can also add some example.

**Comment by wiedzmin** [comment]
@kalbasit btw, whould I finally squash the commits above? Contributing guide seems to say nothing about it, but many other PRs consist of single commit.

File: modules/programs/tmux.nix [hunk 7]
```diff
@@ -277,6 +473,79 @@ in
...
+          ${optionalString (cfg.status.windowStyle != "")
+                           "setw -g window-status-style '${cfg.status.windowStyle}'"}
+          ${optionalString (cfg.status.currentWindowStyle != "")
+                           "setw -g window-status-current-style '${cfg.status.currentWindowStyle}'"}
+          ${optionalString (cfg.status.messageStyle != "")
+                           "setw -g message-style '${cfg.status.messageStyle}'"}
+        '';
+      })
+
+      (mkIf (cfg.status == null) {
```
Comment (rycee) [inline modules/programs/tmux.nix]: Why not simply add these options to the `tmuxConf` variable?

File: modules/programs/tmux.nix [hunk 8]
```diff
@@ -277,6 +473,79 @@ in
...
+          ${optionalString (cfg.status.rightFormat != "")
+                           "set -g status-right '${cfg.status.rightFormat}'"}
+          ${optionalString (cfg.status.style != "")
+                           "set -g status-style '${cfg.status.style}'"}
+          ${optionalString (cfg.status.windowStyle != "")
+                           "setw -g window-status-style '${cfg.status.windowStyle}'"}
+          ${optionalString (cfg.status.currentWindowStyle != "")
+                           "setw -g window-status-current-style '${cfg.status.currentWindowStyle}'"}
+          ${optionalString (cfg.status.messageStyle != "")
+                           "setw -g message-style '${cfg.status.messageStyle}'"}
```
[the remark this answers is not in the document]
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Do you mean all status related options, or just the couple above? I've just grouped together options that are related to status area. The list might be incomplete at the moment, but we have to decide if we have to complete it, or drop entirely. My intent was to have this submodule to emphasize the coupled settings for status area, rather then through them all into extraOptions patchwork.

File: modules/programs/tmux.nix (hunk 5 above)
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Thanks, I was not aware of such solution.

File: modules/programs/tmux.nix [hunk 9]
```diff
@@ -23,14 +53,124 @@ let
...
   defaultResize   = 5;
-  defaultShortcut = "b";
+  defaultShell = "${pkgs.bash}/bin/bash";
+  defaultShortcut = "C-b";
   defaultTerminal = "screen";
 
   boolToStr = value: if value then "on" else "off";
 
   tmuxConf = ''
+    set  -g default-shell "${cfg.shell}"
```
[the remark this answers is not in the document]
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Will it the right thing to change the respective default to "$SHELL"?

File: modules/programs/tmux.nix [hunk 10]
```diff
@@ -8,6 +8,36 @@ let
...
   pluginName = p: if types.package.check p then p.name else p.plugin.name;
 
+  bindingsModule = types.submodule {
+    options = {
+      copyMode = mkOption {
+        type = types.attrs;
+        description = "copy-mode bindings.";
+        default = {};
+      };
+      joinPaneMode = mkOption {
```
[the remark this answers is not in the document]
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Thanks, will do

File: modules/programs/tmux.nix [hunk 11]
```diff
@@ -8,6 +8,36 @@ let
 
   pluginName = p: if types.package.check p then p.name else p.plugin.name;
 
+  bindingsModule = types.submodule {
+    options = {
+      copyMode = mkOption {
+        type = types.attrs;
```
[the remark this answers is not in the document]
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Yep, will use typed attrs.

File: modules/programs/tmux.nix [hunk 12]
```diff
@@ -83,6 +227,14 @@ in
 {
   options = {
     programs.tmux = {
+      activeBorderStyle = mkOption {
```
[the remark this answers is not in the document]
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Thanks, will do

File: modules/programs/tmux.nix (hunk 9 above)
[the remark this answers is not in the document]
Comment (rycee) [reply modules/programs/tmux.nix]: Well I don't use tmux so I don't know what the default value for this option is. I would have expected tmux to use `$SHELL` as the default shell, but your option seems to suggest it is Bash.

My main concern is whether this breaks tmux for zsh users.

File: modules/programs/tmux.nix (hunk 6 above)
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Yes, will do

File: modules/programs/tmux.nix (hunk 7 above)
Comment (wiedzmin) [reply modules/programs/tmux.nix]: I'm thinking of finally move it all to `tmuxConf`, this game of priorities looks awkward.

File: modules/programs/tmux.nix [hunk 13]
```diff
@@ -58,11 +203,10 @@ let
...
     ${optionalString (cfg.shortcut != defaultShortcut) ''
-      # rebind main key: C-${cfg.shortcut}
-      unbind C-${defaultShortcut}
-      set -g prefix C-${cfg.shortcut}
-      bind ${cfg.shortcut} send-prefix
-      bind C-${cfg.shortcut} last-window
+      # rebind main key: ${cfg.shortcut}
+      unbind ${defaultShortcut}
+      set -g prefix ${cfg.shortcut}
+      bind ${cfg.nestedShortcut} send-prefix
```
[the remark this answers is not in the document]
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Yes, will fix

File: modules/programs/tmux.nix (hunk 9 above)
[the remark this answers is not in the document]
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Ok, got it

**Comment by wiedzmin** [comment]
@rycee So we have to decide either to keep `statusSubmodule` and add corresponding test or remove it and fix existing ones. WDYT?

File: modules/programs/tmux.nix [hunk 14]
```diff
@@ -23,14 +51,125 @@ let
...
-  defaultShortcut = "b";
+  defaultShell    = "$SHELL";
+  defaultShortcut = if versionAtLeast config.home.stateVersion "19.09"
+                    then "C-b" else "b";
   defaultTerminal = "screen";
 
   boolToStr = value: if value then "on" else "off";
 
   tmuxConf = ''
+    set  -g default-shell "${cfg.shell}"
```
Comment (kalbasit) [inline modules/programs/tmux.nix]: Let's have `cfg.shell` defaults to null, and include this setting only if it's not null. Personally, I rely on my user's specified shell instead of telling TMux which is my shell.

File: modules/programs/tmux.nix [hunk 15]
```diff
@@ -199,10 +453,10 @@ in
 
       shortcut = mkOption {
         default = defaultShortcut;
-        example = "a";
+        example = "C-a";
```
Comment (kalbasit) [inline modules/programs/tmux.nix]: Should this example be also stateVersion dependent?

Comment (wiedzmin) [reply modules/programs/tmux.nix]: Already done this.

File: modules/programs/tmux.nix (hunk 14 above)
Comment (wiedzmin) [reply modules/programs/tmux.nix]: Ok, makes sense.

**Comment by wiedzmin** [comment]
@kalbasit  @rycee Ok, I'll assume we will keep `statusSubmodule` in place, so I'll push respective test(s) just a bit later.

File: modules/programs/tmux.nix (hunk 5 above)
Comment (rycee) [reply modules/programs/tmux.nix]: Note, the documentation should not depend on the value of `stateVersion` so I believe you'll have to provide a `defaultText` field since the documentation would otherwise be re-built if you change the `stateVersion` value.

File: modules/programs/tmux.nix (hunk 15 above)
Comment (kalbasit) [reply modules/programs/tmux.nix]: Not for the example from what I see.

File: modules/programs/tmux.nix [hunk 16]
```diff
@@ -199,10 +453,10 @@ in
 
       shortcut = mkOption {
         default = defaultShortcut;
```
Comment (rycee) [inline modules/programs/tmux.nix]: Seems to me that this will make the documentation depend on the current configuration. Perhaps use `defaultText` to avoid this?

File: modules/programs/tmux.nix (hunk 14 above)
Comment (rycee) [inline modules/programs/tmux.nix]: I tried installing tmux and reading the documentation. From what I can tell this option is intended for use when tmux is used as a login shell and it should be set to the absolute path of the executable. And by default it will fall back to /bin/sh but I guess only if you don't explicitly set the `default-shell` option?

From what I can tell this is a niche option. I would suggest not attempting to manage this option in Home Manager.

File: modules/programs/tmux.nix [hunk 17]
```diff
@@ -23,14 +51,125 @@ let
     };
   };
 
+  statusModule = types.submodule {
```
Comment (rycee) [inline modules/programs/tmux.nix]: These options need examples.

File: modules/programs/tmux.nix [hunk 18]
```diff
@@ -23,14 +51,125 @@ let
...
+        description = ''
+          Length of right status part.
+        '';
+      };
+
+      style = mkOption {
+        default = "";
+        type = types.str;
+        description = ''
+          Style.
```
Comment (rycee) [inline modules/programs/tmux.nix]: I think this description could be elaborated, I can't tell what it is for. But perhaps it is obvious for an experienced tmux user?

**Comment by rycee** [comment]
I tried putting `programs.tmux.enable = true;` in my configuration and started tmux both without and with this PR applied. Without the PR I get a green bar with information at the bottom of the screen but with the PR I don't. Ideally these changes should be backwards compatible.

**Comment by evanjs** [comment]
Any updates on this? 

Recently have found myself needing something like tmux for remote sessions more often, but I can't seem to get vi bindings working, which this module seems to include...

Would love to see it makes its way into the repo!
